File: examples/eco_route_insight.go
// eco_route_insight.go
//
// A compact insight-economy example: Go computes a local fuel-saving route
// suggestion and emits a small signed envelope, while Python independently
// checks the derivation during testing.
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"eyelingo/internal/exampleinput"
	"fmt"
	"math"
	"sort"
	"time"
)

const eyelingoExampleName = "eco_route_insight"

type Dataset struct {
	CaseName          string           `json:"caseName"`
	Scenario          Scenario         `json:"scenario"`
	Shipment          Shipment         `json:"shipment"`
	CurrentRoute      Route            `json:"currentRoute"`
	AlternativeRoutes []Route          `json:"alternativeRoutes"`
	Policy            Policy           `json:"policy"`
	DataMinimization  DataMinimization `json:"dataMinimization"`
	Signing           Signing          `json:"signing"`
}

type Scenario struct {
	Driver     string `json:"driver"`
	Depot      string `json:"depot"`
	AllowedUse string `json:"allowedUse"`
	IssuedAt   string `json:"issuedAt"`
	TTLHours   int    `json:"ttlHours"`
}

type Shipment struct {
	ID        string  `json:"id"`
	PayloadKg float64 `json:"payloadKg"`
}

type Route struct {
	ID             string  `json:"id"`
	Label          string  `json:"label"`
	DistanceKm     float64 `json:"distanceKm"`
	GradientFactor float64 `json:"gradientFactor"`
	ETAMinutes     int     `json:"etaMinutes"`
}

type Policy struct {
	FuelIndexThreshold               float64 `json:"fuelIndexThreshold"`
	MaxETADelayMinutes               int     `json:"maxEtaDelayMinutes"`
	RequireAlternativeBelowThreshold bool    `json:"requireAlternativeBelowThreshold"`
	AllowedUse                       string  `json:"allowedUse"`
	SignatureAlgorithm               string  `json:"signatureAlgorithm"`
}

type DataMinimization struct {
	ForbiddenEnvelopeTerms []string `json:"forbiddenEnvelopeTerms"`
}

type Signing struct {
	KeyID  string `json:"keyId"`
	Secret string `json:"secret"`
}

type RouteScore struct {
	Route     Route
	FuelIndex float64
	Saving    float64
	Delay     int
	Eligible  bool
}

type Envelope struct {
	Audience   string             `json:"audience"`
	AllowedUse string             `json:"allowedUse"`
	IssuedAt   string             `json:"issuedAt"`
	Expiry     string             `json:"expiry"`
	KeyID      string             `json:"keyId"`
	Assertions EnvelopeAssertions `json:"assertions"`
}

type EnvelopeAssertions struct {
	ShowEcoBanner      bool    `json:"showEcoBanner"`
	SuggestedRoute     string  `json:"suggestedRoute"`
	CurrentFuelIndex   float64 `json:"currentFuelIndex"`
	SuggestedFuelIndex float64 `json:"suggestedFuelIndex"`
	EstimatedSaving    float64 `json:"estimatedSaving"`
	RawDataExported    bool    `json:"rawDataExported"`
}

type Analysis struct {
	PayloadTonnes float64
	CurrentFuel   float64
	Alternatives  []RouteScore
	Selected      RouteScore
	IssueInsight  bool
	Envelope      Envelope
	Canonical     string
	PayloadDigest string
	Signature     string
}

func main() {
	ds := exampleinput.Load(eyelingoExampleName, Dataset{})
	analysis := derive(ds)
	printReport(ds, analysis)
}

func derive(ds Dataset) Analysis {
	payloadTonnes := ds.Shipment.PayloadKg / 1000.0
	currentFuel := fuelIndex(ds.CurrentRoute, payloadTonnes)
	alternatives := make([]RouteScore, 0, len(ds.AlternativeRoutes))
	for _, route := range ds.AlternativeRoutes {
		fi := fuelIndex(route, payloadTonnes)
		saving := currentFuel - fi
		delay := route.ETAMinutes - ds.CurrentRoute.ETAMinutes
		eligible := saving > 0 && delay <= ds.Policy.MaxETADelayMinutes
		if ds.Policy.RequireAlternativeBelowThreshold {
			eligible = eligible && fi <= ds.Policy.FuelIndexThreshold
		}
		alternatives = append(alternatives, RouteScore{Route: route, FuelIndex: fi, Saving: saving, Delay: delay, Eligible: eligible})
	}
	sort.Slice(alternatives, func(i, j int) bool {
		if alternatives[i].Eligible != alternatives[j].Eligible {
			return alternatives[i].Eligible
		}
		if !almostEqual(alternatives[i].Saving, alternatives[j].Saving) {
			return alternatives[i].Saving > alternatives[j].Saving
		}
		return alternatives[i].Route.ID < alternatives[j].Route.ID
	})

	selected := alternatives[0]
	issue := currentFuel > ds.Policy.FuelIndexThreshold && selected.Eligible
	expiry := expiryTime(ds.Scenario.IssuedAt, ds.Scenario.TTLHours)
	envelope := Envelope{
		Audience:   ds.Scenario.Depot,
		AllowedUse: ds.Policy.AllowedUse,
		IssuedAt:   ds.Scenario.IssuedAt,
		Expiry:     expiry,
		KeyID:      ds.Signing.KeyID,
		Assertions: EnvelopeAssertions{
			ShowEcoBanner:      issue,
			SuggestedRoute:     selected.Route.ID,
			CurrentFuelIndex:   round2(currentFuel),
			SuggestedFuelIndex: round2(selected.FuelIndex),
			EstimatedSaving:    round2(selected.Saving),
			RawDataExported:    false,
		},
	}
	canonical := stableJSON(envelope)
	digestBytes := sha256.Sum256([]byte(canonical))
	signature := sign(ds.Signing.Secret, canonical)

	return Analysis{
		PayloadTonnes: payloadTonnes,
		CurrentFuel:   currentFuel,
		Alternatives:  alternatives,
		Selected:      selected,
		IssueInsight:  issue,
		Envelope:      envelope,
		Canonical:     canonical,
		PayloadDigest: hex.EncodeToString(digestBytes[:]),
		Signature:     signature,
	}
}

func fuelIndex(route Route, payloadTonnes float64) float64 {
	return route.DistanceKm * payloadTonnes * route.GradientFactor
}

func expiryTime(issuedAt string, ttlHours int) string {
	parsed, err := time.Parse(time.RFC3339, issuedAt)
	if err != nil {
		panic(err)
	}
	return parsed.Add(time.Duration(ttlHours) * time.Hour).Format(time.RFC3339)
}

func stableJSON(value any) string {
	encoded, err := json.Marshal(value)
	if err != nil {
		panic(err)
	}
	return string(encoded)
}

func sign(secret string, canonical string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(canonical))
	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func round2(x float64) float64 {
	return math.Round(x*100) / 100
}

func almostEqual(a, b float64) bool {
	return math.Abs(a-b) < 0.0000001
}

func yesNo(ok bool) string {
	if ok {
		return "yes"
	}
	return "no"
}

func printReport(ds Dataset, a Analysis) {
	fmt.Println("# Eco Route Insight")
	fmt.Println()
	fmt.Println("## Answer")
	if a.IssueInsight {
		fmt.Println("insight status : issue")
	} else {
		fmt.Println("insight status : hold")
	}
	fmt.Printf("show eco banner : %s\n", yesNo(a.Envelope.Assertions.ShowEcoBanner))
	fmt.Printf("audience : %s\n", a.Envelope.Audience)
	fmt.Printf("allowed use : %s\n", a.Envelope.AllowedUse)
	fmt.Printf("suggested route : %s\n", a.Envelope.Assertions.SuggestedRoute)
	fmt.Printf("current fuel index : %.2f\n", a.Envelope.Assertions.CurrentFuelIndex)
	fmt.Printf("suggested fuel index : %.2f\n", a.Envelope.Assertions.SuggestedFuelIndex)
	fmt.Printf("estimated saving : %.2f\n", a.Envelope.Assertions.EstimatedSaving)
	fmt.Printf("expires at : %s\n", a.Envelope.Expiry)
	fmt.Printf("raw data exported : %s\n", yesNo(a.Envelope.Assertions.RawDataExported))
	fmt.Printf("signature algorithm : %s\n", ds.Policy.SignatureAlgorithm)
	fmt.Printf("payload digest : %s\n", a.PayloadDigest)
	fmt.Printf("signature key : %s\n", a.Envelope.KeyID)
	fmt.Printf("signature : %s\n", a.Signature)
	fmt.Println()
	fmt.Println("## Reason")
	fmt.Println("The current route uses fuel index = distanceKm × (payloadKg / 1000) × gradientFactor.")
	fmt.Printf("For %s, %s gives %.2f × %.2f × %.2f = %.2f.\n", ds.Shipment.ID, ds.CurrentRoute.Label, ds.CurrentRoute.DistanceKm, a.PayloadTonnes, ds.CurrentRoute.GradientFactor, a.CurrentFuel)
	fmt.Printf("The policy threshold is %.2f, so a local eco banner is justified.\n", ds.Policy.FuelIndexThreshold)
	fmt.Printf("The selected alternative %s gives %.2f × %.2f × %.2f = %.2f, saving %.2f while staying within the ETA delay limit.\n", a.Selected.Route.ID, a.Selected.Route.DistanceKm, a.PayloadTonnes, a.Selected.Route.GradientFactor, a.Selected.FuelIndex, a.Selected.Saving)
	fmt.Println("The signed envelope exposes audience, use, expiry, route suggestion, and compact fuel indices, but not raw payload, GPS trace, driver behavior, or raw telemetry.")
	fmt.Println("This follows the insight pattern: ship the decision, not the data.")
	fmt.Println()
}

File: examples/eco_route_insight.js
#!/usr/bin/env node
const crypto = require('crypto');
const { emit, fail, loadInput } = require('./_see');

const NAME = 'eco_route_insight';

function fuelIndex(route, payloadTonnes) {
  return Number(route.distanceKm) * payloadTonnes * Number(route.gradientFactor);
}

function expiry(issuedAt, ttlHours) {
  const date = new Date(issuedAt);
  date.setUTCHours(date.getUTCHours() + Number.parseInt(ttlHours, 10));
  return date.toISOString().replace('.000Z', 'Z');
}

function goNumber(value) {
  const rounded = Math.round((value + Number.EPSILON) * 100) / 100;
  return Number.isInteger(rounded) ? Math.trunc(rounded) : rounded;
}

function chooseAlternative(data, currentFuel) {
  const currentEta = Number.parseInt(data.currentRoute.etaMinutes, 10);
  const maxDelay = Number.parseInt(data.policy.maxEtaDelayMinutes, 10);
  const threshold = Number(data.policy.fuelIndexThreshold);
  const payloadTonnes = Number(data.shipment.payloadKg) / 1000.0;
  const scored = [];
  for (const route of data.alternativeRoutes) {
    const fi = fuelIndex(route, payloadTonnes);
    const saving = currentFuel - fi;
    const delay = Number.parseInt(route.etaMinutes, 10) - currentEta;
    let eligible = saving > 0 && delay <= maxDelay;
    if (data.policy.requireAlternativeBelowThreshold) eligible = eligible && fi <= threshold;
    scored.push({ route, fuelIndex: fi, saving, delay, eligible });
  }
  scored.sort((a, b) => Number(!a.eligible) - Number(!b.eligible) || b.saving - a.saving || a.route.id.localeCompare(b.route.id));
  return scored[0];
}

function canonicalEnvelope(data, currentFuel, selected) {
  const issuedAt = data.scenario.issuedAt;
  const exp = expiry(issuedAt, data.scenario.ttlHours);
  const issue = currentFuel > Number(data.policy.fuelIndexThreshold) && selected.eligible;
  return {
    audience: data.scenario.depot,
    allowedUse: data.policy.allowedUse,
    issuedAt,
    expiry: exp,
    keyId: data.signing.keyId,
    assertions: {
      showEcoBanner: issue,
      suggestedRoute: selected.route.id,
      currentFuelIndex: goNumber(currentFuel),
      suggestedFuelIndex: goNumber(selected.fuelIndex),
      estimatedSaving: goNumber(selected.saving),
      rawDataExported: false,
    },
  };
}

function canonicalJson(value) {
  return JSON.stringify(value);
}

function sign(secret, canonical) {
  return crypto.createHmac('sha256', Buffer.from(secret, 'utf8'))
    .update(Buffer.from(canonical, 'utf8'))
    .digest('base64url');
}

function trustedDerivation(data) {
  const payloadTonnes = Number(data.shipment.payloadKg) / 1000.0;
  const currentFuel = fuelIndex(data.currentRoute, payloadTonnes);
  const selected = chooseAlternative(data, currentFuel);
  const envelope = canonicalEnvelope(data, currentFuel, selected);
  const canonical = canonicalJson(envelope);
  const digest = crypto.createHash('sha256').update(canonical, 'utf8').digest('hex');
  const sig = sign(data.signing.secret, canonical);
  const forbidden = data.dataMinimization.forbiddenEnvelopeTerms;

  fail('Eco route insight derivation failed', {
    'current route crosses fuel threshold': currentFuel > Number(data.policy.fuelIndexThreshold),
    'selected alternative is eligible': selected.eligible,
    'ETA delay is acceptable': selected.delay <= Number.parseInt(data.policy.maxEtaDelayMinutes, 10),
    'selected alternative has best eligible saving': data.alternativeRoutes.every((route) => {
      const fi = fuelIndex(route, payloadTonnes);
      const saving = currentFuel - fi;
      const delay = Number.parseInt(route.etaMinutes, 10) - Number.parseInt(data.currentRoute.etaMinutes, 10);
      const eligible = saving > 0 && delay <= Number.parseInt(data.policy.maxEtaDelayMinutes, 10) && (!data.policy.requireAlternativeBelowThreshold || fi <= Number(data.policy.fuelIndexThreshold));
      return !eligible || selected.saving >= saving;
    }),
    'envelope omits forbidden raw data terms': !forbidden.some((term) => canonical.includes(term)),
    'digest matches canonical envelope': digest === crypto.createHash('sha256').update(canonical, 'utf8').digest('hex'),
    'signature matches canonical envelope': sig === sign(data.signing.secret, canonical),
  });
  return { currentFuel, selected, envelope, digest, sig, canonical };
}

function main() {
  const data = loadInput(NAME);
  const { currentFuel, selected, envelope, digest, sig } = trustedDerivation(data);
  const payloadTonnes = Number(data.shipment.payloadKg) / 1000.0;
  const route = selected.route;
  const saving = selected.saving;
  const status = envelope.assertions.showEcoBanner ? 'issue' : 'clear';
  const yesNo = envelope.assertions.showEcoBanner ? 'yes' : 'no';

  emit('# Eco Route Insight');
  emit();
  emit('## Insight');
  emit(`insight status : ${status}`);
  emit(`show eco banner : ${yesNo}`);
  emit(`audience : ${envelope.audience}`);
  emit(`allowed use : ${envelope.allowedUse}`);
  emit(`suggested route : ${route.id}`);
  emit(`current fuel index : ${currentFuel.toFixed(2)}`);
  emit(`suggested fuel index : ${selected.fuelIndex.toFixed(2)}`);
  emit(`estimated saving : ${saving.toFixed(2)}`);
  emit(`expires at : ${envelope.expiry}`);
  emit('raw data exported : no');
  emit(`signature algorithm : ${data.policy.signatureAlgorithm}`);
  emit(`payload digest : ${digest}`);
  emit(`signature key : ${data.signing.keyId}`);
  emit(`signature : ${sig}`);
  emit();
  emit('## Explanation');
  emit('The current route uses fuel index = distanceKm × (payloadKg / 1000) × gradientFactor.');
  emit(`For ${data.shipment.id}, ${data.currentRoute.label} gives ${Number(data.currentRoute.distanceKm).toFixed(2)} × ${payloadTonnes.toFixed(2)} × ${Number(data.currentRoute.gradientFactor).toFixed(2)} = ${currentFuel.toFixed(2)}.`);
  emit(`The policy threshold is ${Number(data.policy.fuelIndexThreshold).toFixed(2)}, so a local eco banner is justified.`);
  emit(`The selected alternative ${route.id} gives ${Number(route.distanceKm).toFixed(2)} × ${payloadTonnes.toFixed(2)} × ${Number(route.gradientFactor).toFixed(2)} = ${selected.fuelIndex.toFixed(2)}, saving ${saving.toFixed(2)} while staying within the ETA delay limit.`);
  emit('The signed envelope exposes audience, use, expiry, route suggestion, and compact fuel indices, but not raw payload, GPS trace, driver behavior, or raw telemetry.');
  emit('This follows the insight pattern: ship the decision, not the data.');
}

if (require.main === module) main();
